plugins/PostProcessingPlugin/scripts/InsertAtLayerChange.py

```diff
@@ -1,65 +1,218 @@
+                        "every_3rd": "Every 3rd",
+                        "every_5th": "Every 5th",
+                        "every_10th": "Every 10th",
+                        "every_25th": "Every 25th",
+                        "every_50th": "Every 50th",
+                        "every_100th": "Every 100th"},
+                    "default_value": "every_layer",
+                    "enabled": "enable_insatlaychange"
                 },
-                "gcode_to_add":
+                "start_layer":
                 {
-                    "label": "G-code to insert",
-                    "description": "G-code to add before or after layer change.",
-                    "type": "str",
-                    "default_value": ""
+                    "label": "Starting Layer",
+                    "description": "Layer to start the insertion at.  If the Print_Sequence is 'All at Once' then use the layer numbers from the Cura Preview.  Enter '1' to start at gcode LAYER:0. In 'One at a Time' mode use the layer numbers from the first model that prints AND all models will receive the same insertions.  NOTE: There is never an insertion for raft layers.",
+                    "type": "int",
+                    "default_value": 1,
+                    "minimum_value": 1,
+                    "enabled": "insert_frequency != 'once_only' and enable_insatlaychange"
                 },
-                "skip_layers":
+                "end_layer":
                 {
-                    "label": "Skip layers",
-                    "description": "Number of layers to skip between insertions (0 for every layer).",
+                    "label": "Ending Layer",
+                    "description": "Layer to end the insertion at. Enter '-1' to indicate the entire file.  Use layer numbers from the Cura Preview.",
                     "type": "int",
-                    "default_value": 0,
-                    "minimum_value": 0
+                    "default_value": -1,
+                    "minimum_value": -1,
+                    "enabled": "insert_frequency != 'once_only' and enable_insatlaychange"
+                },
+                "single_end_layer":
+                {
+                    "label": "Layer # for Single Insertion.",
+                    "description": "Layer for a single insertion of the Gcode.  Use the layer numbers from the Cura Preview.  The insertion will be at the start of this layer.",
+                    "type": "str",
+                    "default_value": "",
+                    "enabled": "insert_frequency == 'once_only' and enable_insatlaychange"
+                },
+                "gcode_to_add":
+                {
+                    "label": "G-code to insert.",
+                    "description": "G-code to add at start of the layer. Use a comma to delimit multi-line commands. EX: G28 X Y,M220 S100,M117 HELL0.  NOTE:  All inserted text will be converted to upper-case as some firmwares don't understand lower-case.",
```

Comment:
I wouldn't force the upper-case conversion. It is free text, so if your printer doesn't support lower-case, it is up to you to provide only upper-case gcode. On the other hand, if you happen to have a printer that requires lower-case characters for some reason, then you're stuck. Maybe in the note, just advise to take care about it ?

Reply:
Agreed with @rburema that we will keep it as is.